kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/InputFormatReader.java

```diff
@@ -0,0 +1,201 @@
+import org.kitesdk.data.DatasetDescriptor;
+import org.kitesdk.data.DatasetIOException;
+import org.kitesdk.data.DatasetReaderException;
+import org.kitesdk.data.spi.AbstractDatasetReader;
+import org.kitesdk.data.spi.ReaderWriterState;
+
+public class InputFormatReader<E> extends AbstractDatasetReader<E> {
+  private static final TaskAttemptID FAKE_ID =
+      new TaskAttemptID("", 0, false, 0, 0);
+
+  private final FileSystem fs;
+  private final Path path;
+  private final Configuration conf;
+  private final DatasetDescriptor descriptor;
+  private final TaskAttemptContext attemptContext;
+
+  // reader state
+  private ReaderWriterState state = ReaderWriterState.NEW;
+  private Iterator<InputSplit> splits;
+  private RecordReader<E, Void> currentReader = null;
+  private boolean hasNext = false;
+  private boolean shouldAdvance = false;
+
+  public InputFormatReader(FileSystem fs, Path path, DatasetDescriptor descriptor) {
+    this.fs = fs;
+    this.path = path;
+    this.descriptor = descriptor;
+    this.state = ReaderWriterState.NEW;
+
+    // set up the configuration from the descriptor properties
+    this.conf = new Configuration(fs.getConf());
+    for (String prop : descriptor.listProperties()) {
+      conf.set(prop, descriptor.getProperty(prop));
+    }
+
+    this.attemptContext = Hadoop.TaskAttemptContext.ctor.newInstance(conf, FAKE_ID);
+  }
+
+  @Override
+  public void initialize() {
+    Preconditions.checkState(ReaderWriterState.NEW.equals(state),
+        "A reader may not be opened more than once - current state:%s", state);
+
+    try {
+      FileInputFormat format = InputFormatUtil.newInputFormatInstance(descriptor);
+      Job job = Hadoop.Job.newInstance.invoke(conf);
+
+      FileInputFormat.addInputPath(job, path);
+      // attempt to minimize the number of InputSplits
+      FileStatus stat = fs.getFileStatus(path);
+      FileInputFormat.setMaxInputSplitSize(job, stat.getLen());
+
+      this.splits = format.getSplits(job).iterator();
+      this.shouldAdvance = true;
+      this.state = ReaderWriterState.OPEN;
+
+    } catch (RuntimeException e) {
+      this.state = ReaderWriterState.ERROR;
+      throw new DatasetReaderException("Cannot calculate splits", e);
+    } catch (IOException e) {
+      this.state = ReaderWriterState.ERROR;
+      throw new DatasetIOException("Cannot calculate splits", e);
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    Preconditions.checkState(ReaderWriterState.OPEN.equals(state),
+        "Attempt to read from a file in state:%s", state);
+
+    // the Iterator contract requires that calls to hasNext() not change the
+    // iterator state. calling next() should advance the iterator. however,
+    // this wraps a RecordReader that reuses objects, so advancing in next
+    // after retrieving the key/value pair mutates the pair. this hack is a way
+    // to advance once per call to next(), but do it as late as possible.
+    if (shouldAdvance) {
```

Comment:
I think this block needs to be synchronized. Otherwise two threads calling hasNext could advance the state twice.